feat(ESSNTL-4194): Show groups in a table

gkarat opened the pull request:
Implements https://issues.redhat.com/browse/ESSNTL-4194.
Mock https://www.sketch.com/s/c19f555b-7887-4423-b55d-575c8dd1dfb7/a/dl8zwyb.

Create the groups list table with three columns: name, total systems and last modified. Implement sorting by each column, filtering by name and pagination.

⚠️  This doesn't implement actions (rename, delete, create group, bulk select) yet (will be in a separate PR).

## How to test

- Deploy locally with mock server (`npm run start:mock`).
- Navigate to /inventory/groups and check that the table is rendered with dynamically generated values.
- You can also disable the mock server and check whether network errors are properly handled with the error state.
- You can also tweak your mock server and generate empty response (to verify the empty state).
- The table is API paginated and filtered. Any filter change, pagination or sorting you make must trigger a new API request.

## Screenshots

![image](https://user-images.githubusercontent.com/31385370/219063592-3a4f8735-ae74-46f6-b987-c88301334e40.png)

No match state:
![image](https://user-images.githubusercontent.com/31385370/219062552-8f444ca7-7f18-4090-a51b-7e4c70c66b9b.png)

Failed request (any runtime error):
![image](https://user-images.githubusercontent.com/31385370/218801891-d8eb9673-1f1f-4740-940a-4283c2e0f8b3.png)

Pagination values:
![image](https://user-images.githubusercontent.com/31385370/218802132-eae1dce1-e3d2-4904-82ba-ba9a3a2a17bf.png)

computercamplove commented on src/components/GroupsTable/GroupsTable.cy.js:
```diff
@@ -0,0 +1,216 @@
+    descending: 'DESC'
+};
+
+const DEFAULT_ROW_COUNT = 50;
```
@gkarat i remember @mkholjuraev raised a conversation about default pagination 20 - should we stick with it?

Fewwy commented on cypress/support/commands.js:
```diff
@@ -22,4 +22,23 @@
+// Cypress.Commands.overwrite('visit', (originalFn, url, options) => { ... })
+
+Cypress.Commands.add(
+    'ouiaId',
```
I think this command is available in the frontend-component-utils library for cypress
https://github.com/RedHatInsights/frontend-components/blob/master/packages/utils/src/CypressUtils/CustomCommands.js

gkarat replied:
yeah nice

mmenestr commented:
Looks good, just a few comments:
1. Can we reposition the columns so that the "total systems" column is more centered in the table?
2. Can we update the empty state to look like this (just add icon and button) 
<img width="362" alt="image" src="https://user-images.githubusercontent.com/51165119/218807018-71adf4d8-2944-415b-b40f-0a06f8bbf1c4.png">
3. Search should use the "[Search input" component](https://www.patternfly.org/v4/components/search-input) where the icon is on the left
4. For the error empty state, wonder if that action button is really useful (?) is that what we usually use in other places? If so then leave it, but just wondering!

Fewwy commented:
When I tried to change the page I encountered this error

gkarat commented:
@Fewwy,
1. regarding sorting: this is expected, the table sends the `order_by` with the parameter which is not currently listed in the API schema and thus the mock server will throw you error ('name', 'updated_at', `host_ids` are not allowed currently for `order_by` parameter).
2. the same for pagination, the mock server generates crazy values for `count` and `total` fields, thus the requests will be broken too with very huge (forbidden) values.

All these problems will not be reproducible with the real API implementation soon.

Fewwy commented:
@gkarat Other than that the PR look fine :)

gkarat commented:
@mmenestr, thanks a lot for your review! I addressed your points in the last commits. For the button in the error state: this would require some work in our library (separate from the inventory frontend): I left the TODO note to fix it later.

gkarat commented:
/retest

mkholjuraev commented on src/components/GroupsTable/GroupsTable.js:
```diff
@@ -0,0 +1,237 @@
+        setFilters({ ...filters, sortIndex: index, sortDirection: direction });
+    };
+
+    const filterConfigItems = [
```
I am wondering if there is a way to create a more dynamic function to create filters and filter configs... This may be a starting point to improve the filtering on the Inventory table as well.

Now, the SearchInput component is re-created every time on re-render.

gkarat replied:
good point! I think I'll wrap it with useMemo since it looks like a good example where it can become useful

mkholjuraev commented on src/components/GroupsTable/GroupsTable.js:
```diff
@@ -0,0 +1,237 @@
+import { generateLoadingRows } from '../InventoryTable/helpers';
+import NoEntitiesFound from '../InventoryTable/NoEntitiesFound';
+
+const GROUPS_TABLE_INITIAL_STATE = {
```
How about extracting assets into separate file?

gkarat replied:
i'd like to keep it in the same file since these constants are tied to the component's state only

mkholjuraev replied:
Hmm, maybe there will be a new component that can re-use them? But, not a big deal, this is minor :)

mkholjuraev commented on src/components/GroupsTable/GroupsTable.js:
```diff
@@ -0,0 +1,237 @@
+                variant={TableVariant.compact}
+                cells={GROUPS_TABLE_COLUMNS}
+                rows={
+                    uninitialized || loading
```
I think this can be moved to a helper function to have cleaner code. What do you think?

mkholjuraev reviewed (commented):
Codewise looks good!

gkarat commented:
:tada: This PR is included in version 1.3.0 :tada:

The release is available on [GitHub release](https://github.com/RedHatInsights/insights-inventory-frontend/releases/tag/v1.3.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
